src/Language/Fixpoint/Solver/PLE.hs

```diff
@@ -603,9 +619,20 @@ eval γ ctx et = go
     go (PAnd es)        = efAll PAnd (go `traverse` es)
     go (POr es)         = efAll POr (go `traverse` es)
     go e | EVar _ <- dropECst e = do
-      (me', fe) <- evalApp γ ctx e [] et
-      return (Mb.fromMaybe e me', fe)
-    go (ECst e t)       = do (e', fe) <- eval γ ctx et e
+                                    -- If it not on the LHS of a function
+                                    -- application then expand the definition
+                                    -- we need this to not end in an infinite
+                                    -- loop in which we are eta expaning forever
+                                    -- See comment **ETA**, if we are not performing
```

Comment:
The comment **ETA** is rather helpful. Can we also have an word about how the infinite loop would unfold here if not for `isLHSApp`. I still have difficulties to see it.

Reply:
I'll try to explain here and if its ok I'll add as a comment

suppose we have a function f with one argument

We start with the expression `f` when eval is called isLHS is False.
So we end up in the case that is calling eval app on f and we get back `\x -> f x`

Now eval is called on the lambda thus we add in scope x and we call eval on `f x`
We ed up in the case for function application, but before calling evalApp we first call eval on the function and the arguments.

So we end up calling eval on `f` again if we don't have the boolean to tell us that we are on the LHS of a function application we would call again evalApp on `f` alone and obtain again `\x -> f x` then we evaluate the lambda ad so on...

Is it clearer?

Reply:
That helps. I'm still grappling with the argument.

1. `f` is expanded to `\eta1 -> f eta1`. The next time `eval` is called, it is going to find that `f` is applied to all the expected arguments, and therefore `f eta1` is not going to be eta expanded again. Still looks like `isLHSApp` is not needed (?).

2. If `isLHSApp` is kept, `f :: Int -> Int -> Int` is not going to be eta-expanded in
```
C (f x) == C g
```
because it appears in the function position of `f x`. Is this intended?

Reply:
Exactly we get
(\y -> f x) y
Now when we evaluate the lambda inside we can unfold f x
then we perform the replacing of x with y in the expansion

Reply:
There is also the possibility that we don't have the implementation of f but it should be the same thing minus the unfolding

Reply:
How about changing the `EApp` case of `eval` like so
```Haskell
  ...
  (f', fe)  <- case dropECst f of
    EVar _ -> (f, noExpand)
    _ -> eval γ ctx et True f
  (me', fe') <- evalApp γ ctx f' es et
  ...
```
?

This makes the context of the `EVar` handling more explicit and avoids growing the parameters of `eval`.

Reply:
I don't know if it going to solve the problem in general maybe on the lhs we can end up having something like (id f), I remember that I thought about that solution but then I convinced my self it was not enough but I can easily be wrong

Reply:
Corrected